### DISCUSSION TITLE
boltdb fragmentation

### DISCUSSION THREAD

**Discussion by cenkalti** [etcd-io/etcd#15397] [Q&A] [answered]
Does etcd performance impacted by the fragmentation in database?

Is there any benchmarks covering this?

I wonder if I don't defrag the database regularly, am I going to experience any degradation in performance.

**Comment by ptabor** [comment, marked as answer]
Please read: https://github.com/etcd-io/bbolt/issues/401. 

In short: With NoFreelistSync=True, your performance should not be severely impacted... just wested disk space. 
Without NoFreelistSync, your perf will be affected due to need to update list of all free pages in each transaction.

But the cost of NoFreelistSync is paid during DB startup time (as it requires full scan on startup).

**Reply by cenkalti** [reply]
Thanks @ptabor . For reference, etcd sets `NoFreelistSync: true` since 2017.

Commit: https://github.com/etcd-io/etcd/commit/870302afa69b3e10077723b5c33a11044bf5d404

**Reply by ahrtr** [reply]
For small DB size (e.g. < 8G), it doesn't matter what's the value. But indeed it's an issue for large DB size (e.g . > 16G), it may take a long time to open the db in the beginning. 

I think we should expose a flag to let user configure it at least firstly.

**Reply by chaochn47** [reply]
> But indeed it's an issue for large DB size (e.g . > 16G), it may take a long time to open the db in the beginning.

Tested with 10GB DB, it could take 1min22s with the standard AWS GP3 volume. 

It sounds like an opportunity to improve new etcd process bootstrap latency especially when the large old db file is deemed to be replaced with the one transferred from leader but opened unnecessarily.

**Reply by serathius** [reply]
Are you opening 10GB of data on 20GB disk? Have you tried to overprovision it provide more IOPS?
